porch/controllers/rootsyncsets/pkg/controllers/rootsyncset/controller.go

```diff
@@ -202,30 +191,6 @@ func (r *RootSyncSetReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	return ctrl.Result{}, nil
 }
 
-func toCanonicalClusterRef(ref *v1alpha1.ClusterRef, rssNamespace string) (v1alpha1.ClusterRef, error) {
```

Comment:
Are we making all four properties on the clusterRef mandatory in the API here? We use the clusterRef as the key in maps in this controller.

Reply:
I wasn't sure here.  I figured it would still work (?), it might just miss some potential to share caches.  I would like to get this logic into kubebuilder-declarative-pattern - there are a few PRs in flight that support the idea of applying to a remote cluster that we could then just drop in.  I'll try to get those merge-ready today!

I did start introducing a string-clusterKey that e.g. would identify unique clusters based on the URL, but then I spotted that we're not pruning with the canonicalClusterRef (`r.pruneWatches(req.NamespacedName, rootsyncset.Spec.ClusterRefs)`), so I figured the change here was probably more correct anyway?

Happy to upload the clusterKey WIP branch, but I would rather get this into a shared and smaller library, as I think it's useful everywhere and I also think it's quite tricky code and could benefit from being implemented in a minimal standalone library!

Reply:
Moving the logic into a separate library like kubebuilder-declarative-pattern sounds good. The idea behind this was to avoid having to reason about incomplete clusterRefs, we should just make sure they are complete at all times. If we put it back, I can take a pass at simplifying it and make sure it doesn't break the functionality.